projects/igniteui-angular/src/lib/grids/selection/selection.service.ts

```diff
@@ -586,24 +586,36 @@ export class IgxGridSelectionService {
         return this.indeterminateRows.size > 0 && this.indeterminateRows.has(rowID);
     }
 
-    /** Select range from last selected row to the current specified row. */
+    /** Select range from last selected row to the current specified row. */ 
     public selectMultipleRows(rowID, rowData, event?): void {
         this.clearHeaderCBState();
         if (!this.rowSelection.size || this.isRowDeleted(rowID)) {
             this.selectRowById(rowID);
             return;
         }
-        const gridData = this.allData;
-        const lastRowID = this.getSelectedRows()[this.rowSelection.size - 1];
-        const currIndex = gridData.indexOf(this.getRowDataById(lastRowID));
-        const newIndex = gridData.indexOf(rowData);
-        const rows = gridData.slice(Math.min(currIndex, newIndex), Math.max(currIndex, newIndex) + 1);
-        const currSelection = this.getSelectedRowsData();
-        const added = rows.filter(r => !this.isRowSelected(this.getRecordKey(r)));
-        const newSelection = currSelection.concat(added);
-        this.emitRowSelectionEvent(newSelection, added, [], event, currSelection);
-    }
-
+            const gridData = this.allData;
+        const lastSelectedRowID = this.getSelectedRows()[this.rowSelection.size - 1];
+        const currentRowIndex = gridData.findIndex(row => this.getRecordKey(row) === rowID);
+        const lastSelectedRowIndex = gridData.findIndex(row => this.getRecordKey(row) === lastSelectedRowID);
+    
+        const currentGroup = this.getGroup(rowData);
+        const lastSelectedGroup = this.getGroup(this.getRowDataById(lastSelectedRowID));
+    
+        if (!this.areGroupsEqual(currentGroup, lastSelectedGroup)) {
+            this.selectRowById(rowID, false, event);
+            return;
+        }
+    
+        const rangeStart = Math.min(currentRowIndex, lastSelectedRowIndex);
+        const rangeEnd = Math.max(currentRowIndex, lastSelectedRowIndex) + 1;
+        const rowsToSelect = gridData.slice(rangeStart, rangeEnd).filter(row => 
```

Comment:
Generally It seems you grasped what the issue is - the grouped data displayed in the grid does not represent what each row index is in the gridData collection, i.e gridData collection is not the grouped data. The previous code is wrongly assuming `gridData` is what is displayed and is calculating rows to be selected based off on the indexes found in gridData.

I think your fix conception is right - after getting all potential selected rows, find the group they exist in and only select rows from the current group.

I think the code you have written to do this might be optimized. I am pretty sure what you do should exist somewhere as code ( the fact that the group you are trying to define is already displayed, means it has already been resolved/calculated somewhere(). 

Same goes for the `areGroupsEqual` function - most probably we will be reusing some code for that. Still I can give you a feedback - this function is good but it seems to do only a root level comparison of the object - if it has hierarchy in it, this methoud does not considers it.

We will discuss those tomorrow :) Good job so far!